app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart

```diff
@@ -0,0 +1,127 @@
+    final List<FullTask> tasks = await (datastore.queryRecentTasks(slug: slug)).toList();
+
+    // Construct Task columns to scan for backfilling
+    final Map<String, List<FullTask>> taskMap = <String, List<FullTask>>{};
+    for (FullTask fullTask in tasks) {
+      if (taskMap.containsKey(fullTask.task.name)) {
+        taskMap[fullTask.task.name]!.add(fullTask);
+      } else {
+        taskMap[fullTask.task.name!] = <FullTask>[fullTask];
+      }
+    }
+
+    // Check if should be scheduled (there is no yellow runs). Run the most recent gray.
+    final List<Tuple<Target, Task, Commit>> backfill = <Tuple<Target, Task, Commit>>[];
+    for (List<FullTask> taskColumn in taskMap.values) {
+      final FullTask task = taskColumn.first;
+      final CiYaml ciYaml = await scheduler.getCiYaml(task.commit);
+      final Target target = ciYaml.postsubmitTargets.singleWhere((target) => target.value.name == task.task.name);
+      if (target.schedulerPolicy is! BatchPolicy) {
+        continue;
+      }
+      final FullTask? _backfill = _backfillTask(target, taskColumn);
+      if (_backfill != null) {
+        backfill.add(Tuple<Target, Task, Commit>(target, _backfill.task, _backfill.commit));
+      }
+    }
+
+    log.fine('Backfilling ${backfill.length} builds');
+    log.fine(backfill.map<String>((Tuple<Target, Task, Commit> tuple) => tuple.first.value.name));
+
+    // Create list of backfill requests.
+    final List<Future> futures = <Future>[];
+    for (Tuple<Target, Task, Commit> tuple in backfill) {
+      // TODO(chillers): The backfill priority is always going to be low. If this is a ToT task, we should run it at the default priority.
+      final Tuple<Target, Task, int> toBeScheduled = Tuple(
+        tuple.first,
+        tuple.second,
+        LuciBuildService.kBackfillPriority,
+      );
+      futures.add(scheduler.luciBuildService.schedulePostsubmitBuilds(
+        commit: tuple.third,
+        toBeScheduled: [toBeScheduled],
+      ));
+    }
+    // Schedule all builds asynchronously
+    await Future.wait<void>(futures);
+  }
+
+  /// Returns the most recent [FullTask] to backfill.
+  ///
+  /// A [FullTask] is only returned iff:
+  ///   1. There are no running builds (yellow)
+  ///   2. There are tasks that haven't been run (gray)
+  ///
+  /// This is naive, and doesn't rely on knowing the actual Flutter infra capacity.
+  ///
+  /// Otherwise, returns null indicating nothing should be backfilled.
+  FullTask? _backfillTask(Target target, List<FullTask> tasks) {
+    final List<FullTask> relevantTasks = tasks.where((FullTask task) => task.task.name == target.value.name).toList();
+    if (relevantTasks.any((FullTask task) => task.task.status == Task.statusInProgress)) {
+      // Don't schedule more builds where there is already a running task
+      return null;
+    }
+
+    final List<FullTask> backfillTask =
```

Comment:
It would be interesting to use this logic to rerun failed tests to surface flakes. Not needed in this PR though.